### PR TITLE
New Adapter: Seeding Alliance

### PR DESCRIPTION
**Pull request by sag-celejd**
https://github.com/prebid/prebid.github.io/pull/3927

### CODE REVIEW

File: static/bidder-info/seedingAlliance.yaml
Comment (VeronikaSolovei9) [inline static/bidder-info/seedingAlliance.yaml:12]: Please make sure this URL is correct and returns redirect with user macro. 
I tested it with encoded Prebid Server url (https://prebid.adnxs.com/pbs/v1/setuid?bidder=3DseedingAlliance&gdpr=&gdpr_consent=&f=i&uid=$UID): `https://dmp.nativendo.de/set-uuid?gdpr=&gdpr_consent=&redirect_url=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3DseedingAlliance%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%24UID` and expect to see redirect, but see  http 400 error code instead. 

Example of a working URL https://js.adscale.de/pbsync.html?gdpr=&gdpr_consent=&redirect=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3Dbliink%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D
It returns http 200 and then http 302 statuses

File: static/bidder-params/seedingAlliance.json
Comment (VeronikaSolovei9) [inline static/bidder-params/seedingAlliance.json:12]: Just to confirm: adUnitId is not a required parameter and can be omitted, right?

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [inline adapters/seedingAlliance/seedingAlliance.go:10]: While merging with main Prebid Server branch, please set it ti v16.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [inline adapters/seedingAlliance/seedingAlliance.go:17]: Please rename to `type adapter struct {`, there is no need to duplicate a package name here.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [inline adapters/seedingAlliance/seedingAlliance.go:132]: There is no need for this check. Prebid Server gives it to adapter in the same way regardless if imp.ext looks like this 
```
"ext": {
                "seedingAlliance": {
                    "adunitid": "test1"
                }
            }
```
or like this:
```
"ext": {
                "prebid": {
                    "bidder": {
                        "seedingAlliance": {
                            "adunitid": "test1"
                        }
                    }
                }
            }
```

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [inline adapters/seedingAlliance/seedingAlliance.go:46]: Just to confirm: are there any http headers expected?

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [inline adapters/seedingAlliance/seedingAlliance.go:35]: Need to confirm with a team if this is a correct way to set currency like this

**Review by VeronikaSolovei9 — COMMENTED** [review]
Added some comments, please merge your branch with latest Prebid Server master branch

File: static/bidder-params/seedingAlliance.json
Comment (sag-celejd) [reply static/bidder-params/seedingAlliance.json:15]: Thanks for noticing that.  Should have been actually set as "required".

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go:46]: Yes, no headers are expected.

File: static/bidder-info/seedingAlliance.yaml
Comment (sag-celejd) [reply static/bidder-info/seedingAlliance.yaml:12]: The URL itself is fine but we expect a non nil value for **gdpr** on our end.

Meaning that the following URL should work fine:
`https://dmp.nativendo.de/set-uuid?gdpr=0&gdpr_consent=&redirect_url=https://prebid.adnxs.com/pbs/v1/setuid?bidder=seedingAlliance&gdpr=&gdpr_consent=&f=i&uid=$UID`

Is this an issue?

Comment (VeronikaSolovei9) [reply static/bidder-info/seedingAlliance.yaml:12]: Oh I see, please make sure you support encoded redirect URL, this is how PBS will send it.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [reply adapters/seedingAlliance/seedingAlliance.go:35]: Discussed with team. This is ok from our perspective.  For bidder there are can be two "EUR". Do you want to check this array has it already before append?

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [inline adapters/seedingAlliance/seedingAlliance.go:134]: Lines 132-134 are not necessary, I mentioned this in one of my previous comments.

File: static/bidder-info/seedingAlliance.yaml
Comment (sag-celejd) [reply static/bidder-info/seedingAlliance.yaml:12]: Just made sure that we actually support encoded redirect URLs. 👍

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go:134]: Just checked again - you are correct about that during regular usage but our tests fail when we remove line 132-134 since **ext.Bidder** appears to be **nil/empty** when running _TestJsonSamples(t *testing.T)_.

Is this intended?

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go]: Thanks for confirming it!
That's a good point - just pushed a commit with the check.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [reply adapters/seedingAlliance/seedingAlliance.go:142]: You use this format of imp.ext in unite tests: 
```
 "prebid": {
     "bidder": {
          "seedingAlliance": {
                 "adUnitId": "example-tag-id"
           }
       }
}
```
Modify it to 
```
 "prebid": {
       "bidder": {
               "adUnitId": "example-tag-id"
        }
 }

```
Please refer to other bidders examples:
https://github.com/prebid/prebid-server/blob/3b008e46db95a7d31fdfabea975315209feb03c8/adapters/appnexus/appnexustest/exemplary/simple-banner.json#L22

Comment (SyntaxNode) [reply adapters/seedingAlliance/seedingAlliance.go:142]: Bidder parameters are specified in the request at either `imp.ext.prebid.bidders.BIDDER` or the legacy `imp.ext.prebid.BIDDER` locations. PBS moves bidder parameters to `imp.ext.prebid.bidder` for the request. The test cases represent what PBS sends to the bidder code, so the later is the correct format.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (VeronikaSolovei9) [inline adapters/seedingAlliance/seedingAlliance.go:39]: Can you please extract lines 35-40 to a separate function? You can optimize code a little by adding `break` after `eurExists = true`, so it's found already and you don't need to iterate over other elements.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go]: Thank you both, 
`prebid.bidder.adUnitId` didn't work but `ext.bidder.adUnitId` did the trick!

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go]: Done 👍

**Review by VeronikaSolovei9 — COMMENTED** [review]
Thank you for addressing comments. Code looks good to me now. 
I still have http 400 when I test user sync URL, will confirm with a team if it's ok and approve this PR.

File: static/bidder-info/seedingAlliance.yaml
Comment (VeronikaSolovei9) [reply static/bidder-info/seedingAlliance.yaml:12]: We cannot merge this PR if User sync URL doesn't work. It should work with encoded redirect URL and return http 200 and http 302 statuses.

**Comment by VeronikaSolovei9** [comment]
Added comment above about UserSync URL

File: static/bidder-info/seedingAlliance.yaml
Comment (sag-celejd) [reply static/bidder-info/seedingAlliance.yaml:12]: https://dmp.nativendo.de/set-uuid?gdpr=0&gdpr_consent=&redirect_url=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3DseedingAlliance%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%24UID

That 400 response is not coming from our DMP. That status (incl. the following error message: `The bidder name provided is not supported by Prebid Server`) shows up after the redirect since I guess the adapter isn't merged yet.

Comment (SyntaxNode) [reply static/bidder-info/seedingAlliance.yaml:12]: @sag-celejd The user sync url fails if the gdpr field is empty. Prebid Server will set the `gdpr` parameter to either empty, 0, or 1 and all must be supported. An empty gdpr is a lack of a signal versus an explicit 0 or 1 signal. If the gdpr is empty we receive a 400 and thats is what @VeronikaSolovei9 is referring to.

Comment (sag-celejd) [reply static/bidder-info/seedingAlliance.yaml:12]: Alright, I just changed the behaviour on our end:

https://dmp.nativendo.de/set-uuid?gdpr=&gdpr_consent=&redirect_url=https%3A%2F%2Fprebid.adnxs.com%2Fpbs%2Fv1%2Fsetuid%3Fbidder%3DseedingAlliance%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%24UID

Should work fine now.

Comment (VeronikaSolovei9) [reply static/bidder-info/seedingAlliance.yaml:12]: Yes, works good now, thank you for addressing this!

**Review by VeronikaSolovei9 — DISMISSED** [review]
Thank you for addressing changes, LGTM

File: adapters/seedingAlliance/seedingAlliance.go
Comment (mansinahar) [inline adapters/seedingAlliance/seedingAlliance.go:108]: Determining media type of the bid by looking at the media type in the imp is not recommended because in case of multi-format imps, this can lead to an incorrect value depending on how your bidder handles multi-format imps. For example, consider an imp that has both `Native` and `Banner` specified. Now if your bidder returned a `banner` bid then this code will wrongly assign `Native` as the bid type. 

We recommend using `bid.ext.prebid.type` to send this information from your bidder to Prebid Server.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (mansinahar) [inline adapters/seedingAlliance/seedingAlliance.go:113]: Also, I am not sure if these cases for video and audio are needed. The way Prebid Server Core code functions, it won't let through an imp to the adapter if the adapter doesn't support the imp's media type. So unless you have video and audio listed as supported media types in the `seedingAlliance.json` file, this is pretty much non-reachable code. Moreover, in a case where your bidder does receive a request with types other than banner or native, I am assuming it will return an error, no? Again leading to this code being non-reachable.

File: adapters/seedingAlliance/seedingAlliancetest/exemplary/native.json
Comment (mansinahar) [inline adapters/seedingAlliance/seedingAlliancetest/exemplary/native.json:83]: Why is the `expectedBidResponses` empty here? We should be expecting a native bid, no?

File: adapters/seedingAlliance/seedingAlliancetest/supplemental/status_bad_request.json
Comment (mansinahar) [inline adapters/seedingAlliance/seedingAlliancetest/supplemental/status_bad_request.json:16]: Why would this be considered as a bad request?

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go:146]: I added the minLength property - thanks for the hint.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go:108]: That's not a change we would want to do at this point. We are fine with current function and aware of the potential issues. Seems also like that's how many other adapters are doing it.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliance.go]: I am aware that those switch cases should be unreachable code. It's rather a style thing (exhaustive switches etc.).
See my last commit - I just removed the unreachable cases.

File: adapters/seedingAlliance/seedingAlliance.go
Comment (mansinahar) [reply adapters/seedingAlliance/seedingAlliance.go:107]: Sounds good 👍

**Comment by mansinahar** [comment]
@sag-celejd Thanks for addressing the comments. The changes look good. Can you please also check out my comments about the JSON tests and address those when you have a chance?

File: adapters/seedingAlliance/seedingAlliancetest/exemplary/native.json
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliancetest/exemplary/native.json]: Thanks for pointing that out - expectedBidResponses is no longer empty.

File: adapters/seedingAlliance/seedingAlliancetest/supplemental/status_bad_request.json
Comment (sag-celejd) [reply adapters/seedingAlliance/seedingAlliancetest/supplemental/status_bad_request.json:16]: That request itself is trivial/meaningless - I'm just mocking a _Bad Request_ response to test the _MakeBids_ method to accomplish a higher code test coverage.

**Review by mansinahar — APPROVED** [review]

**Review by VeronikaSolovei9 — APPROVED** [review]

**Pull request merged**
